plugins/BEdita/Core/src/Model/Action/SignupUserAction.php

```diff
@@ -0,0 +1,232 @@
+     */
+    public function isValidUrl($value, array $context = [])
+    {
+        // check for a valid scheme (https://, myapp://,...)
+        $regex = '/(?<scheme>^[a-z][a-z0-9+\-.]*:\/\/).*/';
+        if (!preg_match($regex, $value, $matches)) {
+            return false;
+        }
+
+        // if scheme is not an URL protocol then it's a custom url (myapp://) => ok
+        if (!preg_match('/^(https?|ftps?|sftp|file|news|gopher:\/\/)/', $matches['scheme'])) {
+            return true;
+        }
+
+        if (!empty($context['providers']['default'])) {
+            $provider = $context['providers']['default'];
+        } else {
+            $provider = (new Validator())->getProvider('default');
+        }
+
+        return $provider->url($value, true, $context);
+    }
+
+    /**
+     * Create a new user with status draft.
+     *
+     * The user is validated using 'signup' validation.
+     *
+     * @param array $data The data to save
+     * @return mixed
+     */
+    protected function createUser(array $data)
+    {
+        if (!LoggedUser::getUser()) {
+            LoggedUser::setUser(['id' => 1]);
+        }
+
+        $data = ['status' => 'draft'] + $data;
+        $action = new SaveEntityAction(['table' => $this->Users]);
+
+        return $action([
+            'entity' => $this->Users->newEntity(),
+            'data' => $data,
+            'entityOptions' => [
+                'validate' => 'signup',
+            ]
+        ]);
+    }
+
+    /**
+     * Create the signup async job
+     *
+     * @param User $user The user created
+     * @return mixed
+     */
+    protected function createSignupJob(User $user)
+    {
+        $action = new SaveEntityAction(['table' => $this->AsyncJobs]);
+
+        return $action([
+            'entity' => $this->AsyncJobs->newEntity(),
+            'data' => [
+                'service' => 'signup',
+                'payload' => [
+                    'user_id' => $user->id,
+                ],
+                'scheduled_from' => new Time('1 day'),
+                'priority' => 1,
+            ]
+        ]);
+    }
+
+    /**
+     * Send confirmation email to user
+     *
+     * @param User $user The user
+     * @param AsyncJob $job The referred async job
+     * @param array $urlOptions The options used to build activation url
+     * @return void
+     * @throws \Exception When sending email throws an exception different from \Cake\Network\Exception\SocketException
+     */
+    protected function sendMail(User $user, AsyncJob $job, array $urlOptions = [])
+    {
+        try {
+            $options = [
+                'params' => [
+                    'userId' => $user->id,
+                    'activationUrl' => $this->getActivationUrl($job, $urlOptions),
+                ]
+            ];
+            $this->getMailer('BEdita/Core.User')->send('signup', [$options]);
+        } catch (SocketException $e) {
+            // @todo insert async job to retry sending email
```

Comment:
How about writing a custom `AsyncTransport` instead?

On a simple server setup, the default `Mail` or `Smtp` transport would be set, so that emails are sent synchronously.
On a setup that is able to properly manage asynchronous jobs, the custom `Async` transport is set so that every email is not sent immediately, but rather its details are written to the `async_jobs` table, and the job runner will take care to use those very same options on a real transport class (`Mail` or `Smtp` or whatever).

This way, on a multi-instance setup one could avoid telling SMTP secrets to the web servers, and set those pieces of configuration only on workers. 😏